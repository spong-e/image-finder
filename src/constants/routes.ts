const ROUTES = {
  ENTRY: "/",
  DISPLAY: "/display",
  SEARCH: "/search",
};

export { ROUTES };
