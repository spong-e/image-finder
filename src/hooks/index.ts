export { default as useUnsplash } from "./useUnsplash";
