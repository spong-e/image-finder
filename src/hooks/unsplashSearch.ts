import { createApi } from "unsplash-js";

const unsplash = createApi({
  accessKey: process.env.REACT_APP_UNSPLASH_API_KEY ?? "",
});

const unsplashSearch = {
  search: async (topic: string, page: number) => {
    return await unsplash.search.getPhotos({
      query: topic,
      page: page,
      perPage: 1,
    });
  },
};

export { unsplashSearch };
