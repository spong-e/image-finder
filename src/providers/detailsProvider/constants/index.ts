export { ACTION_TYPES } from "./actionTypes";
