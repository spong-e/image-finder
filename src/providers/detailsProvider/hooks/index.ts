export { default as useDetails } from "./useDetails";
export { default as useDetailsActions } from "./useDetailsActions";
