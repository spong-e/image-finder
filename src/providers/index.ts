export { DetailsProvider } from "./DetailsProvider";
export { default as useDetails } from "./useDetails";
export { default as useDetailsActions } from "./useDetailsActions";
