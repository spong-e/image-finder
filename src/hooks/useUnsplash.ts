import { useEffect, useState } from "react";

import { unsplashSearch } from "./unsplashSearch";

const useUnsplash = (topic: string, page: number) => {
  const [isSearching, setIsSearching] = useState<boolean>(false);
  const [photo, setPhoto] = useState<string | null>();
  const [isError, setIsError] = useState<boolean>(false);

  useEffect(() => {
    setIsSearching(true);

    unsplashSearch
      .search(topic, page)
      .then((result) => {
        if (result.type === "success") {
          const photos = result.response;

          const {
            urls: { small },
          } = photos.results[0];

          setPhoto(small);
          setIsSearching(false);
        }
      })
      .catch(() => {
        setIsSearching(false);
        setIsError(true);
        setPhoto(null);
      });

    return () => {};
  }, [page, topic]);

  return [isSearching, photo, isError] as const;
};

export default useUnsplash;
