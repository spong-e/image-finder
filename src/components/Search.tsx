import { FunctionComponent, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { Card, Loader, Message } from "semantic-ui-react";

import { ROUTES } from "../constants";
import { useUnsplash } from "../hooks";
import { useDetails, useDetailsActions } from "../providers";

const Search: FunctionComponent = () => {
  const navigate = useNavigate();
  const details = useDetails();
  const { setDetails } = useDetailsActions();
  const [page, setPage] = useState<number>(1);
  const [isSearching, photo, isError] = useUnsplash(details.topic, page);

  const decline = () => {
    setPage(page + 1);
  };

  const accept = () => {
    const newDetails: Details = { ...details, thumbnail: photo! };
    setDetails(newDetails);
    navigate(ROUTES.DISPLAY);
  };

  if (!details || isError) {
    return (
      <>
        <Message negative>
          <Message.Header>
            Something has gone wrong with your search
          </Message.Header>
          <Link to={ROUTES.ENTRY}>Start again</Link>
        </Message>
      </>
    );
  }

  return (
    <>
      <div>
        {isSearching && (
          <Card.Content>
            <Loader active inline size="massive" data-testid="searchingSpinner">
              Searching...
            </Loader>
          </Card.Content>
        )}
        {photo && (
          <>
            {" "}
            <div>
              <img
                className="preview"
                src={photo}
                alt="preview of search result"
                data-testid="imageElement"
              />
            </div>
            <div>
              <button
                className="button"
                onClick={() => accept()}
                data-testid="acceptBtn"
              >
                Accept
              </button>
              <button
                className="button-decline"
                onClick={() => decline()}
                data-testid="declineBtn"
              >
                Decline
              </button>
            </div>
          </>
        )}
      </div>

      <Link to={ROUTES.ENTRY}>Start again</Link>
    </>
  );
};

export default Search;
