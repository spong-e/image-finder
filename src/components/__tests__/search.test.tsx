import { MemoryRouter } from "react-router-dom";

import { fireEvent, render, screen } from "@testing-library/react";

import { ROUTES } from "../../constants";
import { useUnsplash } from "../../hooks";
import { useDetails, useDetailsActions } from "../../providers";
import Search from "../Search";

import "@testing-library/jest-dom";

const details: Details = {
  firstName: "Test",
  lastName: "User",
  thumbnail: "http://tempuri/cars/1",
  topic: "cars",
};

const mockedNavigate = jest.fn();

const mockedSetDetails = jest.fn();

const mockDetailsActions = {
  setDetails: (details: Details) => mockedSetDetails(details),
};

jest.mock("../../providers");
jest.mock("../../hooks");

jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useNavigate: () => mockedNavigate,
}));

describe.only("Search component", () => {
  describe("Component renders as expected when searching", () => {
    beforeEach(async () => {
      (useDetailsActions as jest.Mock).mockReturnValue(mockDetailsActions);
      (useDetails as jest.Mock).mockReturnValue(details);
      (useUnsplash as jest.Mock).mockImplementation(() => {
        const isSearching = true;
        const photo = null;
        const isError = false;
        return [isSearching, photo, isError] as const;
      });
    });
    it("When component loads, shows 'searching' message and renders as expected", async () => {
      const { container } = render(
        <MemoryRouter>
          <Search />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe("Component renders as expected when error returned", () => {
    beforeEach(async () => {
      (useDetailsActions as jest.Mock).mockReturnValue(mockDetailsActions);
      (useDetails as jest.Mock).mockReturnValue(details);
      (useUnsplash as jest.Mock).mockImplementation(() => {
        const isSearching = false;
        const photo = null;
        const isError = true;
        return [isSearching, photo, isError] as const;
      });
    });
    it("When component loads, shows 'error' message and renders as expected", async () => {
      const { container } = render(
        <MemoryRouter>
          <Search />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe("Component renders as expected when searching complete", () => {
    beforeEach(async () => {
      (useDetailsActions as jest.Mock).mockReturnValue(mockDetailsActions);
      (useDetails as jest.Mock).mockReturnValue(details);
      (useUnsplash as jest.Mock).mockImplementation(
        (topic: string, page: number) => {
          const isSearching = false;
          const photo = `http://tempuri/${topic}/${page}`;
          const isError = false;
          return [isSearching, photo, isError] as const;
        }
      );
    });

    it("When component loads, shows imageu and renders as expected", async () => {
      const { container } = render(
        <MemoryRouter>
          <Search />
        </MemoryRouter>
      );

      const imageWrapper = await screen.findByTestId("imageElement");

      const image = imageWrapper.firstChild as HTMLImageElement;

      expect(image).toHaveAttribute("src", `http://tempuri/${details.topic}/1`);

      expect(container).toMatchSnapshot();
    });

    it("After component loaded, clicking on decline generates a new image", async () => {
      const { container } = render(
        <MemoryRouter>
          <Search />
        </MemoryRouter>
      );

      const imageWrapper = await screen.findByTestId("imageElement");

      const image = imageWrapper.firstChild as HTMLImageElement;

      expect(image).toHaveAttribute("src", `http://tempuri/${details.topic}/1`);

      expect(container).toMatchSnapshot();

      const button = await screen.findByTestId("declineBtn");

      fireEvent.click(button);

      expect(image).toHaveAttribute("src", `http://tempuri/${details.topic}/2`);

      expect(container).toMatchSnapshot();
    });

    it("After component loaded, clicking on accept dispatches action and navigates to display", async () => {
      const { container } = render(
        <MemoryRouter>
          <Search />
        </MemoryRouter>
      );

      const button = await screen.findByTestId("acceptBtn");

      fireEvent.click(button);

      expect(container).toMatchSnapshot();

      expect(mockedSetDetails).toHaveBeenCalledWith(details);

      expect(mockedNavigate).toHaveBeenCalledWith(ROUTES.DISPLAY);
    });
  });
});
