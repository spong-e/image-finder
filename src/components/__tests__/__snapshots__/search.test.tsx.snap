// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component Component renders as expected when error returned When component loads, shows 'error' message and renders as expected 1`] = `
<div>
  <div
    class="ui negative message"
  >
    <div
      class="header"
    >
      Something has gone wrong with your search
    </div>
    <a
      href="/"
    >
      Start again
    </a>
  </div>
</div>
`;

exports[`Search component Component renders as expected when searching When component loads, shows 'searching' message and renders as expected 1`] = `
<div>
  <div>
    <div
      class="content"
    >
      <div
        class="ui massive active text inline loader"
        data-testid="searchingSpinner"
      >
        Searching...
      </div>
    </div>
  </div>
  <a
    href="/"
  >
    Start again
  </a>
</div>
`;

exports[`Search component Component renders as expected when searching complete After component loaded, clicking on accept dispatches action and navigates to display 1`] = `
<div>
  <div>
     
    <div>
      <img
        alt="preview of search result"
        class="preview"
        data-testid="imageElement"
        src="http://tempuri/cars/1"
      />
    </div>
    <div>
      <button
        class="button"
        data-testid="acceptBtn"
      >
        Accept
      </button>
      <button
        class="button-decline"
        data-testid="declineBtn"
      >
        Decline
      </button>
    </div>
  </div>
  <a
    href="/"
  >
    Start again
  </a>
</div>
`;

exports[`Search component Component renders as expected when searching complete After component loaded, clicking on decline generates a new image 1`] = `
<div>
  <div>
     
    <div>
      <img
        alt="preview of search result"
        class="preview"
        data-testid="imageElement"
        src="http://tempuri/cars/1"
      />
    </div>
    <div>
      <button
        class="button"
        data-testid="acceptBtn"
      >
        Accept
      </button>
      <button
        class="button-decline"
        data-testid="declineBtn"
      >
        Decline
      </button>
    </div>
  </div>
  <a
    href="/"
  >
    Start again
  </a>
</div>
`;

exports[`Search component Component renders as expected when searching complete After component loaded, clicking on decline generates a new image 2`] = `
<div>
  <div>
     
    <div>
      <img
        alt="preview of search result"
        class="preview"
        data-testid="imageElement"
        src="http://tempuri/cars/2"
      />
    </div>
    <div>
      <button
        class="button"
        data-testid="acceptBtn"
      >
        Accept
      </button>
      <button
        class="button-decline"
        data-testid="declineBtn"
      >
        Decline
      </button>
    </div>
  </div>
  <a
    href="/"
  >
    Start again
  </a>
</div>
`;

exports[`Search component Component renders as expected when searching complete When component loads, shows imageu and renders as expected 1`] = `
<div>
  <div>
     
    <div>
      <img
        alt="preview of search result"
        class="preview"
        data-testid="imageElement"
        src="http://tempuri/cars/1"
      />
    </div>
    <div>
      <button
        class="button"
        data-testid="acceptBtn"
      >
        Accept
      </button>
      <button
        class="button-decline"
        data-testid="declineBtn"
      >
        Decline
      </button>
    </div>
  </div>
  <a
    href="/"
  >
    Start again
  </a>
</div>
`;
